### DISCUSSION TITLE
How to crop based on biggest shape in photo - PHP & JS

### DISCUSSION THREAD

**Discussion by Whaoo** [ImageMagick/ImageMagick#5335] [Help]
Hello friends, 

I'm trying to crop an image based on the biggest "object" in the picture, which will always be a paper sheet with text. 

In "easy" photos, imagecropauto from GD is easy enought and works like a charm. 

But on more difficult images, nothings works, tryed despeckleImage, trim, imagecropauto. 

[Result of B&W processing](https://prnt.sc/wOLb4Nl2VADF)
[Result of B&W processing reverted](https://prnt.sc/OZIvRa8lsLXv)

I tried imageconvolution based on 3x3 array [Ridge detection](https://en.wikipedia.org/wiki/Ridge_detection) and Sharpen. 

But I can't get rid of the lines and other "shapes". 

The wanted result is the following : 

[Wanted result (hope you enjoy my paint skills)](https://prnt.sc/3qK-PwO-lbpS) 

I strongly believed what I want can be achieved via imageMagick

Any ideas ? ❤️ 

Saw : https://github.com/ImageMagick/ImageMagick/discussions/2002 & https://github.com/ImageMagick/ImageMagick/discussions/4039 but I need it on PHP no command lines 😢

**Comment by snibgo** [comment]
You don't show any before-and-after images, such as an input and the corresponding required output. So I don't know what you really mean by:
> crop an image based on the biggest "object" ...

However, I don't use PHP, so probably can't help.

**Reply by Whaoo** [reply]
Hello,

Sorry, just a newbie. 

So here is what i'm at : 

`$im = imagecreatefromjpeg('test.jpg');`
[An exemple of image ](https://prnt.sc/JYW2s0apn2ey)

```
imagefilter($im, IMG_FILTER_GRAYSCALE);
imagefilter($im, IMG_FILTER_CONTRAST, -1000);
imagejpeg($im, "test_B&W.jpg");
```

[Black and white image - test_B&W.jpg](https://prnt.sc/qPlM9n32HgQD)

```
imagefilter($im, IMG_FILTER_NEGATE);
imagejpeg($im, "test_B&W_reversed.jpg");
```
[Black and white image - test_B&W.jpg](https://prnt.sc/UvVqvrKE9vbP)

Here the B&W is I think more exploitable. 

I then use : 
```
$cropped = imagecropauto($im , IMG_CROP_WHITE);
if ($cropped !== false) { 
    imagedestroy($im);  
    $im = $cropped;     
}
imagejpeg($im, "test_B&W_cropped.jpg");
```
 [Black and white image cropped - test_B&W_cropped.jpg](https://prnt.sc/OmwpIlXplDcX)

I then tried using GD and the IMG_CROP_WHITE, IMG_CROP_BLACK, IMG_CROP_DEFAULT without luck. 

And 

```
$imagick = new \Imagick(realpath("./test_B_W_cropped.jpg"));
$imagick ->trimImage(-1);
$imagick ->setImageBackgroundColor(new \ImagickPixel('transparent'));
header("Content-Type: image/" . $imagick ->getImageFormat());
echo $imagick ;
```
[output of trim ](https://prnt.sc/NhfBI7TYYdLx)


And that's it, I'm stuck. 

- The wanted result are either : 

Get the coordinates and size of the square shape to then crop with specific pixels points. 
OR
Crop directly based on biggest "square" shape from photo. 

Like :[(auto detect this shape)](https://prnt.sc/v50EgqzfB14Q)
[And the crop and display FINAL RESULT  ](https://prnt.sc/PO4hO67tv4B5)

**Comment by snibgo** [comment]
Okay. The problem isn't trivial. The image ("someText.png") has only three corners. Ideally, we would have four corners, and could use a perspective distortion. But we can manage with just three.

Windows BAT syntax. Sorry, I don't do PHP.
```
magick ^
  someText.png ^
  -distort Perspective "190,434 0,0 774,157 700,0 1135,725 700,700" ^
  -crop 700x700+0+0 +repage ^
  out.png
```
out.png is:
![out](https://user-images.githubusercontent.com/33812368/179796259-e1169785-f407-4802-900c-ce91c54fcb70.png)

I found the coordinates of the corners by using an image viewer. But we need to find them automatically. We can do that by simplifying the image get get white paper on a black background (and with black text). Then `-canny` finds the edges, and `-hough-lines` makes those edges into straight lines.
```
magick ^
  someText.png ^
  -alpha off ^
  -channel RGB ^
  -auto-level ^
  +channel ^
  -threshold 75%% ^
+write x1.png ^
  -fill Black +opaque White ^
+write x2.png ^
  -canny 0x10+10%%+15%% ^
+write x3.png ^
  -hough-lines 9x9+100 ^
+write x4.png ^
  x.mvg
```
Each "+write" is just so we can see what is happening.

x3.png is:
![x3](https://user-images.githubusercontent.com/33812368/179796467-66569ad7-4c24-4104-b5c7-90b3b9287349.png)

x4.png shows the major lines, which intersect at the corners of the paper:
![x4](https://user-images.githubusercontent.com/33812368/179796624-605373b9-874f-4972-86f0-ce7d725ef5b5.png)

x.mvg is a text file:
```
# Hough line transform: 9x9+100
viewbox 0 0 1295 970
# x1,y1 x2,y2 # count angle distance
line 669.38,0 1299.31,970  # 253 147 633
line 0,519.365 1295,-84.5034  # 319 65 673
line 12.0444,0 403.95,970  # 355 158 1323
line 0,1380.18 1295,632.513  # 197 60 1367

```

A shell script (or PHP) can read x.mvg, and calculate the intersections. Or it might read the angles (147 degrees, etc) and rotate the image accordingly.

**Reply by Whaoo** [reply]
Oh wow, that's more than impressive. 😱 

I never used Win BAT, but i'll try my best to understand fully the code provided then try to transpose it to php/exec command line. 

The only questions i have in mind currently are (sorry they may sound stupid) : 

1. When you use canny and hough-lines (that I didn't know of thanks a lot for the discovery) :  

`-canny 0x10+10%%+15%% ^`
`-hough-lines 9x9+100 ^`
Could you help me undersand the parameters? (ints) used ? 
For canny i looked and it's for (-canny radiusxsigma{+lower-percent}{+upper-percent}) but how do you choose the values ?

2. If the image is compsed of 2 elements, like 2 tickets, 1 with at least 3 borders, 1 with only two. Can it automaticly choose and focus the crop on the "at least 3 borders" element ? 

[Photo exemple ](https://prnt.sc/_sE6BMeQzaCt)

**Reply by himynameisgeoff** [reply]
Hi, sorry. This thread seems to be what I'm looking for. Could you explain how to retrieve the coordinates from x.mvg and pipe them into a batch file for use in cropping?

**Reply by snibgo** [reply]
Your question is vague. x.mvg is a text file. It can be read and processed by shell commands.

Your question is in the middle of a thread. I suggest you start a new thread, explaining what you want to do, and what help you need.

**Comment by snibgo** [comment]
For all operations, see:
https://imagemagick.org/script/command-line-options.php

For `-canny` see:
https://www.imagemagick.org/discourse-server/viewtopic.php?f=4&t=25405
https://www.imagemagick.org/Usage/transform/#canny

For `-hough-lines` see:
https://www.imagemagick.org/discourse-server/viewtopic.php?f=4&t=25476
https://www.imagemagick.org/Usage/transform/#hough

If `-deskew` worked for this image, it would greatly simplify my solution. Sadly, it doesn't. (At least, I can't make it work.)

> If the image is compsed of 2 elements, like 2 tickets, 1 with at least 3 borders, 1 with only two. Can it automaticly choose and focus the crop on the "at least 3 borders" element ?

Yes, but it needs more processing. If I run that image through my command, x2.png is:
![x2](https://user-images.githubusercontent.com/33812368/179806104-a41a8ebe-a43c-4fc1-94f8-c4ac72e6a647.png)

We can use `-connected-components` to find the largest white area. Then we can paint all other areas black. This will eliminate the smaller paper.

Alternatively, we might iterate `-hough-lines`, reducing the "count" parameter (100 in my example above) until we have just four lines.

For your next difficult example, you might have overlapping papers. That would be more difficult to solve.

**Reply by Whaoo** [reply]
I'll take a look at all those links.  ❤️ and come back with hopefully a working php/command lines code 

Yes haha next issue will be overlapping ones. But first, i'll go throught the first two cases

**Reply by Whaoo** [reply]
Hello @snibgo ,

Just had the approval to used exec() in my code.

So first I applied the code you provided.

My x3 seems to be more detailed and x4 represents multiple lines and not only 4.

My params are off ?

[X4.png](https://prnt.sc/H5LbnvERcLGt)

Edit: Ok, I've figured out the issue partially which is the export on prnt seems to alter the original image. 

Here is the original jpg image : https://www.swisstransfer.com/d/25a25247-05b2-453e-98ab-1af990b813cc

But even fixing 200 ou 0x1 i still have way too many lines :/

**Comment by snibgo** [comment]
Your test.jpg (4640x3472) is about 3.6 times the size of your OP image, which I call someText.png (1295x970).

To get the same effect, we need to increase the `sigma` parameter of the `-canny` operation, and all parameters of `-hough-lines`. We multiply them by 3.6.
```
  -canny 0x36+10%%+15%% ^

  -hough-lines 32x32+360 ^
```
This fixes the problem, for that test.jpg.

You probably want a general solution for all image sizes. We can use `%[fx:...]` expressions. test.jpg has width 4640. 
```
  -canny 0x%%[fx:w*36/4640]+10%%+15%% ^

  -hough-lines %%[fx:w*32/4640]x%%[fx:w*32/4640]+%%[fx:w*360/4640] ^
```
For Windows BAT, we need to double each %.

This now works for both test.jpg and someText.png.

**Reply by Whaoo** [reply]
😮 OK that's impressive. 

Yes for the png - jpg previous exports. I did the rookie thing to just make a screenshot with auto upload. 

I end up with only 3 hough-lines with the one dedicated to test.jpg, do I need to extrapolate or did I again missed something ? 


Code:
`
exec('magick test.jpg -alpha off -channel RGB -auto-level +channel -threshold 75%% +write x1.png -fill Black +opaque White +write x2.png -canny 0x36+10%%+15%% +write x3.png -hough-lines 32x32+360 +write x4.png x.mvg');
`

And the general solution creates the x1 & x2 but no x3, x4. Maybe I made a mistake and the exec() from php do not need the dubble %%, but if it wors with the code above... 🤨 

`exec('magick test.jpg -alpha off -channel RGB -auto-level +channel -threshold 75%% +write x1.png -fill Black +opaque White +write x2.png -canny 0x%%[fx:w*36/4640]+10%%+15%%  +write x3.png -hough-lines %%[fx:w*32/4640]x%%[fx:w*32/4640]+%%[fx:w*360/4640] +write x4.png x.mvg');` 

I start looking for my mistakes tomorrow morning. Thanks a lot ❤️

**Comment by snibgo** [comment]
> I end up with only 3 hough-lines with the one dedicated to test.jpg, ...

Ah, yes, I didn't notice that. My eyesight isn't what it used to be. Tweaking `-hough-lines` fixes that (says he hopefully, squinting closely at the screen).
```
  -hough-lines %%[fx:w*60/4640]x%%[fx:w*60/4640]+%%[fx:w*350/4640] ^
```

Doubled `%%` is needed for Windows BAT. I don't know PHP, sorry. I suggest you read PHP documentation, and what escapes are needed.

In your PHP, I suggest you check for any errors raised by `magick`.

As a general point: the `%[fx:...]` expressions are to cope with inputs of different sizes. Large images take much longer to process than small images. So an alternative strategy is to always resize inputs to a fairly small and constant size, eg `-resize 1000x1000` so images fit in a 1000x1000 box. This will simplify processing, and make it faster, but reduces the precision of the intersections (which would need to be scaled up, of course). A reduced precision may not be a problem.

**Reply by Whaoo** [reply]
Thanks a lot ! 

I have difficulties collecting the errors from imagisk, will try understand all that during the weekend and come back monday with working code and maybe questions for improvement  🤞
